source/include/core_mqtt.h

```diff
@@ -101,6 +101,68 @@ typedef void (* MQTTEventCallback_t )( struct MQTTContext * pContext,
                                        struct MQTTPacketInfo * pPacketInfo,
                                        struct MQTTDeserializedInfo * pDeserializedInfo );
 
+/**
+ * @brief User defined API used to store outgoing publishes. Used to track any publish 
+ * retransmit on an unclean session connection.
+ *
+ * @param[in] pContext Initialised MQTT Context.
+ * @param[in] packetId Outgoing publish packet identifier.
+ * @param[in] pIoVec Pointer to the outgoing publish packet in form of array of Tansport Vectors.
+ * @param[in] ioVecCount Number of transport vectors in the pIoVec array.
+ *
+ * @return True if the copy is successful else false.
+ */
+/* @[define_mqtt_retransmitstorepacket] */
+typedef bool ( * MQTTRetransmitStorePacket)( struct MQTTContext * pContext,
+                                            uint16_t packetId,
+                                            TransportOutVector_t * pIoVec,
+                                            size_t ioVecCount );
+/* @[define_mqtt_retransmitstorepacket] */
+
+/**
+ * @brief User defined API used to retreive a copied publish for resend operation. Used to 
+ * track any publish retransmit on an unclean session connection.
+ *
+ * @param[in] pContext Initialised MQTT Context.
+ * @param[in] packetId Copied publish packet identifier.
+ * @param[in] pIoVec Output parameter to store the pointer to the copied publish packet form of array of Tansport Vectors.
+ * @param[in] ioVecCount Output parameter to store the number of transport vectors in the pIoVec array.
+ *
+ * @return True if the retreive is successful else false.
+ */
+/* @[define_mqtt_retransmitretrievepacket] */
+typedef bool ( * MQTTRetransmitRetrievePacket)( struct MQTTContext * pContext,
```

Comment:
I feel that, provided we are giving them the `TransportOutVector_t` array to copy while storing, if we make them store the packet in a contiguous memory then it will be more difficult for them to manage their heap. This is because:
1. it will have untidy way to search for a packet ID during retrieval if they store the packet ID within the same memory.
2. while clearing a packet they will have to either leave an empty block which is not guaranteed to be fit for storing a new packet or they will have to move the other packets back to fill in the empty space (which is only possible if the packet ID is stored within the same memory else we cannot move the data is it will tamper with the pointers that are meant to store the data)
3. as can be seen from point 2, big chunks of fragmented memory might still remain.

If stored in the form of `TransportOutVector_t` array itself, it will be easier for them to manage their heap. Easy search, easy storing, tidy clearing.

Hence, even though both the methods lead to similar levels of fragmentation, the method of storing in form of `TransportOutVector_t` array is tidier. 

I have created an example for the way the users can write these callbacks in a tidy manner.

```
#define MAX_PUBLISH_COUNT 3U

#define MAX_SIZE_OF_PUBLISH_PACKET 64U

#define MAX_PUBLISH_VECTOR_COUNT 4U

typedef struct publishCopy
{
    uint16_t packetId;

    uint8_t buffer[ MAX_SIZE_OF_PUBLISH_PACKET ];

    TransportOutVector_t pIoVec[ MAX_PUBLISH_VECTOR_COUNT ];

} publishCopy_t;

static publishCopy_t publishCopyBuffer[ MAX_PUBLISH_COUNT ];

bool publishStoreCallback(struct MQTTContext* pContext,
                        uint16_t packetId,
                        TransportOutVector_t* pIoVec,
                        size_t ioVecCount)
{
    bool found = false;

    for ( int i = 0; i < MAX_PUBLISH_COUNT; i++ )
    {
        if (publishCopyBuffer[i].packetId == 0)
        {
            uint8_t* bufferPointer = publishCopyBuffer[i].buffer;

            publishCopyBuffer[i].packetId = packetId;

            for (int j = 0; j < ioVecCount; j++)
            {
                memcpy( ( void* ) bufferPointer, ( const void* ) pIoVec[j].iov_base, pIoVec[j].iov_len);

                publishCopyBuffer[i].pIoVec[j].iov_base = bufferPointer;
                publishCopyBuffer[i].pIoVec[j].iov_len = pIoVec[j].iov_len;

                bufferPointer += pIoVec[j].iov_len;
            }

            found = true;

            break;
        }
    }

    return found;
}

bool publishRetrieveCallback(struct MQTTContext* pContext,
                            uint16_t packetId,
                            TransportOutVector_t** pIoVec,
                            size_t* ioVecCount)
{
    bool found = false;

    for ( int i = 0; i < MAX_PUBLISH_COUNT; i++ )
    {
        if (publishCopyBuffer[i].packetId == packetId)
        {
            *pIoVec = publishCopyBuffer[i].pIoVec;
            *ioVecCount = 4;

            found = true;

            break;
        }
    }

    return found;
}

bool publishClearCallback(struct MQTTContext* pContext,
                          uint16_t packetId)
{
    bool found = false;

    for (int i = 0; i < MAX_PUBLISH_COUNT; i++)
    {
        if (publishCopyBuffer[i].packetId == packetId)
        {
            memset(&publishCopyBuffer[i], 0, sizeof(publishCopy_t));

            found = true;

            break;
        }
    }

    return found;
}
```

Reply:
I think it makes sense as allows more flexibility in storage. I'd probably use contiguous memory, but can do that by just passing a single elem iovec on return and concatenating the elems on receiving.

Though if nobody would use the flexibility in storage, its just additional overhead for all users to deal with.
For saving a packet, I can see it being useful though as the MQTT library should not need the memory for concatenating the segments; this way the user can concat them directly into the destination buffer. For retrieving a packet, I'm not sure if I see much value other than consistency.

Reply:
Thank you for the answer @DakshitBabbar,

The method suggested above seems a bit inefficient in terms of memory. What I mean by that is the user must allocate arrays of fixed size. That leads to inefficient use of memory which is quite an important resource on embedded systems. Any packet which is even 2 bytes long will take up whole slot which can used by others.

For an example, let us consider a scenario where the outstanding packet is limited to 2 and the array size is 1000. So, 2 packets of size 500 and 1500 cannot be sent. Even though there is memory.

Also, the method suggested above has more memory requirement only to store the data-structure. It is storing array of arrays.

Also, when the user just wants to call malloc/free to store and clear a value they will need to call malloc and free on many vectors one at a time. It will cause very small memory allocations and deallocations which causes fragmentation.

What I am proposing is this:

```
typedef struct userPublishStore {
    uint8_t *buffer;
    size_t bufferlen;
    uint16_t packetId;
} userPublishStore_t;

userPublishStore_t storedPublishes[MAX_PUBLISH_COUNT];

bool publishStoreCallback(
    struct MQTTContext *pContext,
    uint16_t packetId,
    TransportOutVector_t *pIoVec,
    size_t ioVecCount
) {
    bool ret = false;
    size_t memory_required = 0;
    size_t store_location;

    for (store_location = 0; store_location < MAX_PUBLISH_COUNT;
         store_location++) {
        if (storedPublishes[store_location].packetId == 0) {
            break;
        }
    }

    if (store_location == MAX_PUBLISH_COUNT) {
        return false;
    }

    for (size_t i = 0; i < ioVecCount; i++) {
        memory_required += pIoVec[i].iov_len;
    }

    // The following is user dependent memory management function.
    // They can implement it however they want. They can simply use malloc
    // or get it from an array or anything else.
    uint8_t *buffer = user_memory_management_get_function(memory_required);
    size_t index = 0;

    for (size_t i = 0; i < ioVecCount; i++) {
        memcpy(&buffer[index], pIoVec[i].iov_base, pIoVec[i].iov_len);
        index += pIoVec[i].iov_len;
    }

    storedPublishes[store_location].buffer = buffer;
    storedPublishes[store_location].packetId = packetId;
    storedPublishes[store_location].bufferlen = memory_required;
    return true;
}

bool publishRetrieveCallback(struct MQTTContext* pContext,
                            uint16_t packetId,
                            uint8_t * buffer,
                            size_t * bufferLength)
{
    bool found = false;

    for ( int i = 0; i < MAX_PUBLISH_COUNT; i++ )
    {
        if (storedPublishes[i].packetId == packetId)
        {
            // No need to memcpy here. The buffer is on "user-heap" and stays there unless freed.
            buffer = storedPublishes[i].buffer;
            *bufferLength = storedPublishes[i].bufferlen;

            found = true;

            break;
        }
    }

    return found;
}

bool publishClearCallback(struct MQTTContext* pContext,
                          uint16_t packetId)
{
    bool found = false;

    for (int i = 0; i < MAX_PUBLISH_COUNT; i++)
    {
        if (storedPublishes[i].packetId == packetId)
        {
            // User is free to choose how to "free" this memory. They can simply call free or memset it to 0 or any other choice.
            user_memory_management_free_function(storedPublishes[i].buffer);
            storedPublishes[i].packetId = 0;
            found = true;
            break;
        }
    }

    return found;
}
```

Please Ignore the flouting of some MISRA rules in my example and inconsistent naming style :)

But, do go over the above and let me know if you think there are some flows with the system.

This is a blocker for me as we are now defining an API - the internal workings can change but I do not want to change any API often.

Thanks